File: python/algorithms/greedy_empty_set.py
import networkx as nx
from algorithms.compute_PC import compute_pc
from algorithms.local_search import local_search_procedure
from utils.improvement_condition import improvement_condition
from tqdm import tqdm

def greedy_empty_set(
    G: nx.Graph, terminals: list[int], 
    k: int, case: 1 | 2) -> tuple[set, list[int]]:
  
  # If graph is too big, use sparse representation
  # for better memory usage

  T : set = set(terminals)
  V : set = set(G.nodes)

  S : set = set()
  # pc_deltas : list[int] = []

  K : set = set()
  
  if case == 1:
    K = set()

  elif case == 2:
    K = set(T)

  while len(K) != len(V) - k:

    best_j = None
    best_pc = float("inf")

    for j in V - K:
      
      S_j = V - (K | {j})
      pc_j = compute_pc(G, S_j, terminal_nodes=terminals)

      # print(pc_j)
      
      # argmin
      if pc_j < best_pc:
        best_pc = pc_j
        best_j = j

    K.add(best_j)
    # pc_deltas.append(best_pc)

  S = V - K

  return S

def extended_critical_node_empty_set(
    G, terminals, k, case, maxIter, use_ls: False):
  
  S = None
  best_S = None
  best_pc = float("inf")
  
  # for _ in tqdm(range(maxIter), desc="greedy ES", total=maxIter):

  curr_S = None

  S = greedy_empty_set(G, terminals=terminals, k=k, case=case)

  curr_S = set(S)

  if use_ls:
    S_ls = local_search_procedure(
      G, S, 
      terminals=terminals, 
      case=case, 
      improvement_condition=improvement_condition)
    
    curr_S = set(S_ls)
  
  curr_pc = compute_pc(G, curr_S, terminal_nodes=terminals)

    # if curr_pc < best_pc:
    #   best_pc = curr_pc
    #   best_S = set(curr_S)
    #   print(f"Refined PC: {best_pc}\n")

  return curr_S, curr_pc
  # return best_S, best_pc

File: c_env/algorithms/greedy_empty_set.cpp
#include "greedy_empty_set.hpp"

#include <limits>

#include "improvement_condition.hpp"
#include "local_search.hpp"
#include "utils.hpp"

std::unordered_set<int> greedy_empty_set(
    const GraphData &graph_data,
    const std::vector<int> &terminals,
    int k,
    int case_value)
{
  (void)terminals;

  std::unordered_set<int> T(
      terminals.begin(),
      terminals.end());

  std::unordered_set<int> V(
      graph_data.nodes.begin(),
      graph_data.nodes.end());

  std::unordered_set<int> S;
  std::unordered_set<int> K;

  if (case_value == 1)
  {
    K.clear();
  }
  else if (case_value == 2)
  {
    K = T;
  }

  while (static_cast<int>(K.size()) != static_cast<int>(V.size()) - k)
  {
    int best_j = -1;
    int best_pc = std::numeric_limits<int>::max();

    for (int j : V)
    {
      if (K.count(j))
      {
        continue;
      }

      std::unordered_set<int> candidate_kept = K;
      candidate_kept.insert(j);

      std::unordered_set<int> S_j;
      for (int node : V)
      {
        if (!candidate_kept.count(node))
        {
          S_j.insert(node);
        }
      }

      int pc_j = compute_pc(graph_data, S_j, terminals);

      if (pc_j < best_pc)
      {
        best_pc = pc_j;
        best_j = j;
      }
    }

    if (best_j == -1)
    {
      break;
    }

    K.insert(best_j);
  }

  for (int node : V)
  {
    if (!K.count(node))
    {
      S.insert(node);
    }
  }

  return S;
}

std::pair<std::unordered_set<int>, int> extended_critical_node_empty_set(
    const GraphData &graph_data,
    const std::vector<int> &terminals,
    int k,
    int case_value,
    int max_iter,
    bool use_local_search)
{
  (void)max_iter;

  std::unordered_set<int> S;
  std::unordered_set<int> current_S;
  int best_pc = std::numeric_limits<int>::max();

  S = greedy_empty_set(graph_data, terminals, k, case_value);
  current_S = S;

  if (use_local_search)
  {
    std::unordered_set<int> S_ls = local_search_procedure(
        graph_data,
        S,
        terminals,
        case_value,
        improvement_condition);
    current_S = S_ls;
  }

  int current_pc = compute_pc(graph_data, current_S, terminals);
  return {current_S, current_pc};
}